### ISSUE TITLE
model.evaluate generates different accuracy and loss values on the same network and test dataset 

### ISSUE THREAD

**Issue by hamidkhb** [tensorflow/tensorflow#39602]
System Info:
Windows 10 
Tensorflow 2.1.0
Im using GPU Nvidia v100-sxm2 on a cluster. 

I have trained a CNN model using Tensorflow 2 for gesture classification. The training is done and the results are good. my problem is for my task (Im trying to use Reinforcement Learning to prune the network) I need to evaluate the network many times. I have not changed the testing dataset. and didnt change anything about the network between evaluating. I have tried to freeze the network `model.trainable = False `  I thought it might have an influece but didn't help. 

I know it's not an issue but I already asked in [stackoverflow](https://stackoverflow.com/questions/61822122/running-model-evaluate-many-times-results-different-accuracy-and-loss-value-tens) but didn't get any useful answer and it effects my job alot. 

my code:

making dataset:

```
def split_dataset(dataset: tf.data.Dataset, validation_data_fraction: float):

    validation_data_percent = round(validation_data_fraction * 100)
    if not (0 <= validation_data_percent <= 100):
        raise ValueError("validation data fraction must be ∈ [0,1]")

    dataset = dataset.enumerate()
    train_dataset = dataset.filter(lambda f, data: f % 100 >= validation_data_percent)
    validation_dataset = dataset.filter(lambda f, data: f % 100 < validation_data_percent)

    # remove enumeration
    train_dataset = train_dataset.map(lambda f, data: data)
    validation_dataset = validation_dataset.map(lambda f, data: data)

    return train_dataset, validation_dataset

def load_data(path):
    data, label = data_prep(path)
    dataset = tf.data.Dataset.from_tensor_slices((data, label))
    dataset = dataset.shuffle(100000)
    train_dataset, rest = split_dataset(dataset, 0.3)
    test_dataset, valid_dataset = split_dataset(rest, 0.5)
    train_data = train_dataset.shuffle(1000).batch(10)
    valid_data = valid_dataset.batch(10)
    test_data = test_dataset.batch(10)
    return train_data, valid_data, test_data
```

my model:

```
def make_model():
    model = tf.keras.Sequential([
    Input((1,30,30)),
    Conv2D(filters = 8, kernel_size=(3,3), padding="same", activation="relu", name="c1", data_format="channels_first"),
    Conv2D(filters = 16, kernel_size=(3,3), padding="same", activation="relu", name="c2", data_format="channels_first"),
    MaxPool2D(pool_size=(2,2), strides=(1,1),padding="same", name="m1", data_format="channels_first"),
    
    Conv2D(filters = 16, kernel_size=(3,3), padding="same", activation="relu", name="c3", data_format="channels_first"),
    MaxPool2D(pool_size=(2,2), strides=(1,1),padding="same", name="m2",data_format="channels_first"),
    
    Flatten(),
    Dense(256, activation="relu", use_bias=True),
    Dense(5,  use_bias=True)])
    return model


model = make_model()
print(model.summary())
model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=0.001),
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True), metrics=["accuracy"])
model.fit(train_data, verbose=1, validation_data=valid_data, epochs=20)
```

and evaluating:

```
model.evaluate(test_data)

```

**Comment by deadshotsb** [comment]
Shuffling the data and splitting it to train and validation set is a good way to make both the sets more uniformly distributed. Though a random shuffle does not always guarantees the uniform distribution always. There may be cases where some data is present in the validation set but not present in the training set, hence changing the accuracy of the model. The changes in accuracy are however very minute.

**Comment by hamidkhb** [comment]
@deadshotsb Thats true. But the thing is even if all the weights are random and the data is bad it shouldn't change the output of the network if I evaluate the same data N times in a row. The problem is not the performance of the network, it's actually working very well, but the fact that it changes. I'm not able to understand this small randomnes in the output. here is for example the result of 5 times running `model.evaluate (test_data) in a row:

 885/Unknown - 2s 2ms/step - loss: 0.1039 - accuracy: 0.9663
885/Unknown - 2s 2ms/step - loss: 0.0959 - accuracy: 0.9675
885/Unknown - 2s 2ms/step - loss: 0.0999 - accuracy: 0.9661
885/Unknown - 2s 2ms/step - loss: 0.0888 - accuracy: 0.9688
885/Unknown - 2s 2ms/step - loss: 0.0799 - accuracy: 0.9715

**Comment by deadshotsb** [comment]
@hamidkhb Consider that your data contains samples of A (95%) and B (5%). But when we are randomly arranging the dataset into the train and validation, a case may occur that only few samples of that 5% of B goes to training set and majority of B sample goes to validation. So in this case, I don't expect my model to perform well in predicting of sample B. Whereas if the opposite case occurs that I have maximum of B samples in my training set, then it is expected to perform well in predicting sample B. Generally, it occurs due to huge difference between each kind of samples in the dataset.

**Comment by hamidkhb** [comment]
@deadshotsb I understand this and this is true. But it still doesn't explain the random output of the test. the test dataset contains specific number of images that aren't changing every time I use them to evaluate the same network the output is different. Bad distribution of datasets can't explain this. putting aside that the network is performing very well on unseen data (the results are in my previous post).

**Comment by hamidkhb** [comment]
I found the Problem. apparantely running tf.data.Dataset changes the order of the data in the dataset. It's working somehow as a random iterator and does that every time one uses it. thats why it's generating a bit of randomnes

**Comment by deadshotsb** [comment]
yup told you

**Comment by hamidkhb** [comment]
@deadshotsb It wasn't caused by the distribution of the data. Like mentioned before the performance of the network is very well. It was caused by different order of the **same** data in the iterator of test dataset. This caused different mean value of loss and accuracy. evaluating the same data in test dataset **manually** solved the problem.
